NuoDB/0.0.9-0.0.A.sql

```diff
@@ -0,0 +1,325 @@
+   AND UPPER(does_exist.host) = UPPER(inHost)
+   AND ((UPPER(does_exist.plus) = UPPER(inPlus)) OR (does_exist.plus IS NULL AND inPlus IS NULL))
+  WHERE does_exist.id IS NULL
+ );
+END_IF;
+RETURN (
+ SELECT id
+ FROM Email
+ WHERE UPPER(username) = UPPER(inUserName)
+  AND UPPER(host) = UPPER(inHost)
+  AND ((UPPER(plus) = UPPER(inPlus)) OR (plus IS NULL AND inPlus IS NULL))
+);
+
+END_FUNCTION;
+@
+SET DELIMITER ;
+
+DROP FUNCTION IF EXISTS GetEmail/1;
+SET DELIMITER @
+CREATE OR REPLACE FUNCTION GetEmail (
+ inEmail STRING
+) RETURNS INTEGER AS
+
+VAR plus_part STRING;
+VAR user_part STRING = (SELECT SUBSTRING_INDEX(inEmail, '@', 1) FROM DUAL);
+VAR host_part STRING = (SELECT SUBSTRING_INDEX(inEmail, '@', -1) FROM DUAL);
+VAR plus_idx INTEGER = (SELECT LOCATE('+', user_part) FROM DUAL);
+
+IF (plus_idx > 0)
+ plus_part = (SELECT SUBSTR(user_part, plus_idx + 1) FROM DUAL);
+ IF (plus_part = '')
+  plus_part = NULL;
+ END_IF;
+ user_part = (SELECT SUBSTR(user_part, 1, plus_idx - 1) FROM DUAL);
+END_IF;
+
+RETURN GetEmail(user_part, plus_part, host_part);
+
+END_FUNCTION;
+@
+SET DELIMITER ;
+
+DROP FUNCTION IF EXISTS GetListIndividualName;
+SET DELIMITER @
+CREATE OR REPLACE FUNCTION GetListIndividualName (
+ inListName STRING,
+ inSetName STRING
+) RETURNS INTEGER AS
+VAR listName_id INTEGER;
+VAR setName_id INTEGER;
+VAR listIndividual_id INTEGER;
+
+IF (inListName IS NOT NULL)
+ -- Get names
+ listName_id = GetWord(inListName);
+ setName_id = GetWord(inSetName);
+
+ -- Insert list name if it does not exist
+ INSERT INTO ListIndividualName (name, listSet, optinStyle)
+ SELECT listName_id, setName_id, 1
+ FROM DUAL
+ LEFT JOIN ListIndividualName AS does_exist ON does_exist.name = listName_id
+  AND ((does_exist.listSet = setName_id) OR (does_exist.listSet IS NULL AND setName_id IS NULL))
+  AND optinStyle = 1
+ WHERE does_exist.listIndividual IS NULL
+ LIMIT 1
+ ;
+END_IF;
+
+-- Get individual list
+RETURN (
+ SELECT listIndividual
+ FROM ListIndividualName
+ WHERE name = listName_id
+  AND ((listSet = setName_id) OR (listSet IS NULL AND setName_id IS NULL))
+  AND optinStyle = 1
+ LIMIT 1
+);
+END_FUNCTION;
+@
+SET DELIMITER ;
+
+DROP FUNCTION IF EXISTS ListSubscribe/4;
+SET DELIMITER @
+CREATE OR REPLACE FUNCTION ListSubscribe (
+ inListName STRING,
+ inSetName STRING,
+ inIndividual BIGINT,
+ inSend STRING
+) RETURNS INTEGER AS
+VAR listIndividual_id INTEGER;
+VAR sendField_id INTEGER;
+
+IF (inIndividual IS NOT NULL AND inListName IS NOT NULL)
+ sendField_id = GetIdentifier(LOWER(inSend));
+ listIndividual_id = GetListIndividualName(inListName, inSetName);
+
+ -- Insert individual into list
+ INSERT INTO ListIndividual (id, individual, type)
+ SELECT listIndividual_id AS id, inIndividual AS individual, sendField_id AS type
+ FROM DUAL
+ LEFT JOIN ListIndividual AS does_exist ON does_exist.id = listIndividual_id
+  AND does_exist.individual = inIndividual
+  AND does_exist.unlist IS NULL
+ WHERE does_exist.id IS NULL
+ LIMIT 1
+ ;
+END_IF;
```

Comment:
Do we need to set listIndividual_id again after the insert so that it returns the id of the inserted ListIndividual if insert was necessary?

Reply:
Never mind